bot/exts/moderation/defcon.py

```diff
@@ -28,71 +35,81 @@
 
 BASE_CHANNEL_TOPIC = "Python Discord Defense Mechanism"
 
+SECONDS_IN_DAY = 86400
+
 
 class Action(Enum):
     """Defcon Action."""
 
-    ActionInfo = namedtuple('LogInfoDetails', ['icon', 'color', 'template'])
+    ActionInfo = namedtuple('LogInfoDetails', ['icon', 'emoji', 'color', 'template'])
 
-    ENABLED = ActionInfo(Icons.defcon_enabled, Colours.soft_green, "**Days:** {days}\n\n")
-    DISABLED = ActionInfo(Icons.defcon_disabled, Colours.soft_red, "")
-    UPDATED = ActionInfo(Icons.defcon_updated, Colour.blurple(), "**Days:** {days}\n\n")
+    SERVER_OPEN = ActionInfo(Icons.defcon_unshutdown, Emojis.defcon_unshutdown, Colours.soft_green, "")
+    SERVER_SHUTDOWN = ActionInfo(Icons.defcon_shutdown, Emojis.defcon_shutdown, Colours.soft_red, "")
+    DURATION_UPDATE = ActionInfo(
+        Icons.defcon_update, Emojis.defcon_update, Colour.blurple(), "**Threshold:** {threshold}\n\n"
+    )
 
 
 class Defcon(Cog):
     """Time-sensitive server defense mechanisms."""
 
-    days = None  # type: timedelta
-    enabled = False  # type: bool
+    # RedisCache[str, str]
+    # The cache's keys are "threshold" and "expiry".
+    # The caches' values are strings formatted as valid input to the DurationDelta converter, or empty when off.
+    defcon_settings = RedisCache()
 
     def __init__(self, bot: Bot):
         self.bot = bot
         self.channel = None
-        self.days = timedelta(days=0)
+        self.threshold = relativedelta(days=0)
+        self.expiry = None
+
+        self.scheduler = Scheduler(self.__class__.__name__)
 
-        self.bot.loop.create_task(self.sync_settings())
+        self.bot.loop.create_task(self._sync_settings())
 
     @property
     def mod_log(self) -> ModLog:
         """Get currently loaded ModLog cog instance."""
         return self.bot.get_cog("ModLog")
 
-    async def sync_settings(self) -> None:
+    @defcon_settings.atomic_transaction
+    async def _sync_settings(self) -> None:
         """On cog load, try to synchronize DEFCON settings to the API."""
+        log.trace("Waiting for the guild to become available before syncing.")
         await self.bot.wait_until_guild_available()
         self.channel = await self.bot.fetch_channel(Channels.defcon)
 
-        try:
-            response = await self.bot.api_client.get('bot/bot-settings/defcon')
-            data = response['data']
+        log.trace("Syncing settings.")
 
-        except Exception:  # Yikes!
+        try:
+            settings = await self.defcon_settings.to_dict()
+            self.threshold = parse_duration_string(settings["threshold"]) if settings["threshold"] else None
```

Comment:
Can we avoid the startup KeyError on first use by using `.get()` here?
```suggestion
            self.threshold = parse_duration_string(settings["threshold"]) if settings.get("threshold") else None
```

Reply:
Is it possible that we won't be able to read from redis properly? because if there is, a KeyError will allow us to find out about it.

Reply:
I'm not aware of such a situation, if we had connection problems then that'd manifest elsewhere.

Reply:
Right, but then you are also reminded that your defcon settings aren't synced and that you should bring them back if defcon was active. If it's not desirable I don't mind changing it, but that's pretty much the goal of the try clause.

Reply:
I think the proper solution here is to have a fallback that this cog falls to, not just let everything crash and burn. If we want to add an alert system, that's fine, but it's complementary to the other solution.

Reply:
It does have fallback, it just sets to the default 0. You can still use defcon, it's just unsynced

Reply:
I mean we shouldn't error out just because redis is having trouble. The cog should continue without errors, or with proper error handling, and the report should be handled seperately.

Reply:
It continues, it just lets you know that it wasn't able to sync

Reply:
I can suppress any errors while updating actually. But other than that defcon is functional even without a connection.

Reply:
@jb3 yeah so in short the purpose here is to catch that key error or any other error that might indicate things aren't working well. Do you want to remove that?

Reply:
So, I haven't tested with redis locally, but if we start up the bot and don't immediately set a defcon will it just alert the moderators on every startup?

I think a better idea is to alert and then actually take action and set the preferences.

Reply:
No, it will only misfire when the PR is merged because the cache is new. Once we set the cache once it won't do that anymore.

Reply:
Practically speaking you run `!dc t 0` once and that's it. The issue is that you need to have the keys in the cache.